### PR TITLE
Cleanup and refactoring for gridmanager

### PR DESCRIPTION
**Pull request by halungge**
Clean up of the single node `grid_manager.py`:

- `GridManager` now implements the context manager protocol 
- added additional global property names for grid files of different origin (MPI-M, DWD...)
- clean up of the `GridFile`: to contain simple read functions for the different netcdf elements
- more consistend usage of `GridFile` througout `GridManager`
- converted static methods to simple module functions

added introduction of `refinement.py`:
- simple functionality for handling what we need so far from the grid refinement.

### CODE REVIEW

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
launch jenkins spack

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
launch jenkins spack

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 1]
```diff
@@ -26,495 +36,662 @@ def __init__(self, *args, **kwargs):
...
+        If a index array is given it only reads the values at those positions.
+        Args:
+            name: name of the field to read
+            indices: indices to read
+            dtype: datatype of the field
+        """
+        try:
+            variable = self._dataset.variables[name]
+            _log.debug(f"reading {name}: {variable}")
+            data = variable[:] if indices is None else variable[indices]
```
[the remark this answers is not in the document]
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/grid_manager.py:288]: There was something that that was needed, netcdf has a bit a strange interface, but Ill try again.

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 2]
```diff
@@ -26,495 +36,662 @@ def __init__(self, *args, **kwargs):
...
-         e2    e1
-        /    c0  \
-        ----e0----
-        \   c1   /
-         e3    e4
-          \   /
-           \ /
-
-        For example, for this diamond for e0 -> (e1, e2, e3, e4)
+def _update_size_for_1d_sparse_dims(grid):
```
[the remark this answers is not in the document]
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/grid_manager.py:582]: good catch, thanks

File: model/common/src/icon4py/model/common/grid/grid_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/grid_manager.py:288]: Indeed, this is necessary in order to get the data.

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 3]
```diff
@@ -26,495 +36,662 @@ def __init__(self, *args, **kwargs):
...
+            return data
         except KeyError as err:
             msg = f"{name} does not exist in dataset"
-            self._log.warning(msg)
+            _log.warning(msg)
+            _log.debug(f"Error: {err}")
             raise exceptions.IconGridError(msg) from err
 
+    def close(self):
+        self._dataset.close()
```
[the remark this answers is not in the document]
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/grid_manager.py:298]: @static_method is generally discouraged, and this one is not even static? It accesses class instance state...

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
launch jenkins spack

**Review by nfarabullini — APPROVED** [review]

**Comment by halungge** [comment]
launch jenkins spack

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 4]
```diff
@@ -26,495 +36,663 @@ def __init__(self, *args, **kwargs):
...
+        Args:
+            name: name of the field to read
+            transpose: flag to indicate whether the file should be transposed (for 2d fields)
+        Returns:
+            xp.ndarray: field data
+
+        """
+        _log.debug(f"reading {name}: transposing = {transpose}")
+        data = self.variable(name, indices, dtype=gtx.int32)
+        return xp.transpose(data) if transpose else data
```
Comment (OngChia) [inline model/common/src/icon4py/model/common/grid/grid_manager.py:272]: Do you know whether the data is on the GPU device when read from netCDF? 
If the data is not on the device, I think we need to add `xp.asarray(xp.transpose(data)) if transpose else xp.asarray(data)` to return an object on the device. I suppose the connectivity needs to be on the device.

File: model/common/src/icon4py/model/common/grid/refinement.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+TODO: Are these the x_grf dimension in the netcdf grid file?
+"""
+
+
+_UNORDERED: Final[dict[gtx.Dimension : tuple[int, int]]] = {
+    dims.CellDim: (0, -4),
+    dims.EdgeDim: (0, -8),
+    dims.VertexDim: (0, -4),
+}
+"""Value indicating a point is int the unordered interior (fully prognostic) region: this is encoded by 0 or -4 in coarser parent grid."""
```
Comment (OngChia) [inline model/common/src/icon4py/model/common/grid/refinement.py:51]: "is in"

**Review by OngChia — CHANGES_REQUESTED** [review]

**Review by OngChia — APPROVED** [review]

**Comment by halungge** [comment]
launch jenkins spack

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 6]
```diff
@@ -26,495 +36,663 @@ def __init__(self, *args, **kwargs):
...
-        C2V = "vertex_of_cell"  # does not exist in grid.simple.py
+class FieldName(GridFileName):
+    ...
 
-        #: name of E2C connectivity in grid file: dims(nc=2, edge)
-        E2C = "adjacent_cell_of_edge"
 
-        #: name of C2E connectivity in grid file: dims(nv=3, cell)
-        C2E = "edge_of_cell"
+class ConnectivityName(FieldName):
```
Comment (OngChia) [inline model/common/src/icon4py/model/common/grid/grid_manager.py:156]: Sorry for another late comment. Is it good to have two separate classes such as IndicesFieldName for all connectivities and FiedlName for other float varaibles that indicate coordinates etc?

**Comment by halungge** [comment]
cscs-ci run default

File: model/common/src/icon4py/model/common/grid/grid_manager.py [hunk 7]
```diff
@@ -26,495 +36,663 @@ def __init__(self, *args, **kwargs):
...
+    END_INDEX_VERTICES = "end_idx_v"
+
+
+class GridFile:
+    """Represent and ICON netcdf grid file."""
+
+    INVALID_INDEX = -1
+
+    def __init__(self, file_name: str):
+        self._filename = file_name
```
[the remark this answers is not in the document]
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/grid_manager.py:246]: Ui, indeed that would be much better...

File: model/common/src/icon4py/model/common/grid/grid_manager.py (hunk 6 above)
Comment (halungge) [reply model/common/src/icon4py/model/common/grid/grid_manager.py:156]: This is more a semantic oriented decision really not technically based on the netcdf structure. For netcdf those are just fields.

**Comment by halungge** [comment]
cscs-ci run default

**Comment by halungge** [comment]
launch jenkins spack

**Comment by halungge** [comment]
cscs-ci run default

**Pull request merged**